### PR TITLE
Cart server refactor

### PR DESCRIPTION
**Pull request by joshsadam**
## Description of changes

Cleaning up server call to better facilitate getting the files for the cart.

## Related issue

Link to the GitHub issue this pull request addresses using the `#issuenum` format. If it completes an issue, use `Fixes #issuenum` to automatically close the issue.

## Checklist

Things for the developer to confirm they've done before the PR should be accepted:

*   [x] CHANGELOG.md (and UPGRADING.md if necessary) updated with information for new change.
*   [x] Tests added (or description of how to test) for any new features.
*   [ ] ~User documentation updated for UI or technical changes.~

### CODE REVIEW

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/ajax/dto/cart/CartProject.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package ca.corefacility.bioinformatics.irida.ria.web.ajax.dto.cart;
+
+import java.util.List;
+
+import ca.corefacility.bioinformatics.irida.ria.web.ajax.dto.CartSample;
+
+/**
+ * Model for UI to represent a project and its sample within the cart.
+ */
+public class CartProject {
```
Comment (deepsidhu85) [inline src/main/java/ca/corefacility/bioinformatics/irida/ria/web/ajax/dto/cart/CartProject.java:10]: As this is a model for the UI rename it to `CartProjectModel`

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/cart/CartController.java [hunk 2]
```diff
@@ -1,49 +1,32 @@
...
-	private final SampleService sampleService;
-	private final ProjectService projectService;
-
 	@Value("${server.base.url}")
 	private String serverBaseUrl;
 	/*
 	 * Additional variables
 	 */
-	private String iridaPipelinePluginStyle;
+	private final String iridaPipelinePluginStyle;
```
Comment (deepsidhu85) [inline src/main/java/ca/corefacility/bioinformatics/irida/ria/web/cart/CartController.java:25]: Remove the `final`

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/cart/CartGalaxyController.java [hunk 3]
```diff
@@ -12,27 +13,26 @@
...
 /**
  * Controller to handle all ajax requests made to the cart that have to do with Galaxy.
  */
 @RestController
-@Scope("session")
 @RequestMapping("/ajax/galaxy-export")
 public class CartGalaxyController {
-	private SampleService sampleService;
-	private Cart cart;
+	private final SampleService sampleService;
```
Comment (deepsidhu85) [inline src/main/java/ca/corefacility/bioinformatics/irida/ria/web/cart/CartGalaxyController.java:29]: Remove the `final` from the variables

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/cart/CartGalaxyController.java [hunk 4]
```diff
@@ -41,14 +41,9 @@ public CartGalaxyController(SampleService sampleService, Cart cart) {
...
 	public List<GalaxyExportSample> getGalaxyExportForm() {
-		Map<Long, Map<Long, CartSample>> contents = cart.get();
-		List<GalaxyExportSample> result = new ArrayList<>();
-		for (Long projectId : contents.keySet()) {
-			Iterable<Sample> samples = sampleService.readMultiple(contents.get(projectId)
-					.keySet());
-			samples.forEach(s -> result.add(new GalaxyExportSample(s, projectId)));
-		}
-		return result;
+		Map<Project, List<Sample>> contents = cartService.getFullCart();
```
Comment (deepsidhu85) [inline src/main/java/ca/corefacility/bioinformatics/irida/ria/web/cart/CartGalaxyController.java:44]: Might need to apply the irida intellij styling to this method

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/pipelines/PipelineController.java [hunk 5]
```diff
@@ -83,45 +83,41 @@
...
 	 * SERVICES
 	 */
-	private ReferenceFileService referenceFileService;
-	private SequencingObjectService sequencingObjectService;
-	private AnalysisSubmissionService analysisSubmissionService;
-	private ProjectService projectService;
-	private UserService userService;
-	private IridaWorkflowsService workflowsService;
-	private MessageSource messageSource;
+	private final ReferenceFileService referenceFileService;
```
Comment (deepsidhu85) [inline src/main/java/ca/corefacility/bioinformatics/irida/ria/web/pipelines/PipelineController.java:86]: Remove the `final` from the variables

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/pipelines/PipelineController.java [hunk 6]
```diff
@@ -83,45 +83,41 @@
...
+	private final EmailController emailController;
+	private final UICartService cartService;
 
 	@Autowired
 	public PipelineController(SequencingObjectService sequencingObjectService,
 			ReferenceFileService referenceFileService, AnalysisSubmissionService analysisSubmissionService,
 			IridaWorkflowsService iridaWorkflowsService, ProjectService projectService, UserService userService,
-			CartController cartController, MessageSource messageSource,
+			MessageSource messageSource,
 			final WorkflowNamedParametersService namedParameterService, UpdateSamplePermission updateSamplePermission,
```
Comment (deepsidhu85) [inline src/main/java/ca/corefacility/bioinformatics/irida/ria/web/pipelines/PipelineController.java:105]: Remove `final`

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/services/UICartService.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+		}
+		return cartProjects;
+	}
+
+	/**
+	 * Get the entire cart flushed out into {@link Project}s with their {@link Sample}s
+	 *
+	 * @return All proejcts and samples in the part
+	 */
+	public Map<Project, List<Sample>> getFullCart() {
```
Comment (deepsidhu85) [inline src/main/java/ca/corefacility/bioinformatics/irida/ria/web/services/UICartService.java:116]: What do you think about changing wherever you have `Map<Project, List<Sample>>` here and any of the above files to a dto which contains those objects?

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/sessionAttrs/Cart.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Session object to hold samples that are currently in the cart.
+ */
+public class Cart extends HashMap<Long, HashSet<Long>> {
+
+	public int add(Long projectId, List<Long> sampleIds) {
```
Comment (deepsidhu85) [inline src/main/java/ca/corefacility/bioinformatics/irida/ria/web/sessionAttrs/Cart.java:13]: Method comment

File: src/main/webapp/resources/js/apis/cart/cart.js [hunk 9]
```diff
@@ -1,47 +1,49 @@
...
 
 /**
  * Get the current state of the cart.
  * @returns {Promise<void | never>}
  */
 export const getCart = async () =>
-  axios.get(`${url}`).then(response => response.data);
+  axios.get(`${AJAX_URL}`).then((response) => response.data);
 
 export const getCartIds = async () =>
```
Comment (deepsidhu85) [inline src/main/webapp/resources/js/apis/cart/cart.js:35]: Add comments to functions

File: src/main/webapp/resources/js/pages/cart/components/CartSamples.jsx [hunk 10]
```diff
@@ -72,18 +72,27 @@ function CartSamplesComponent({
...
+  removeProject,
 }) {
-  const filterSamples = e => applyFilter(e.target.value);
+  const s = samples
+    .map((project) => {
+      const { samples, ...p } = project;
+      return project.samples.map((sample) => ({ ...sample, project: p }));
+    })
+    .flat();
+  console.log(s);
```
Comment (deepsidhu85) [inline src/main/webapp/resources/js/pages/cart/components/CartSamples.jsx:83]: Remove this logging statement and also change the const variable name to make it clear as to what it is

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]
Working good just some changes

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/ajax/CartAjaxController.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+	 *
+	 * @return list of project identifiers
+	 */
+	@GetMapping("/ids")
+	public ResponseEntity<Set<Long>> getProjectIdsInCart() {
+		return ResponseEntity.ok(service.getProjectIdsInCart());
+	}
+
+	/**
+	 * Cart the samples that are in the cart for a specific project
```
[the remark this answers is not in the document]
Comment (joshsadam) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/ajax/CartAjaxController.java:90]: Fixed in 7ff3cc60b7ad101a19b06bf2ea94e6e007a93528

File: CHANGELOG.md [hunk 12]
```diff
@@ -6,6 +6,7 @@ Changes
 * [UI]: Fixed bug where sequencing runs could not be deleted on sequencing runs details page. (20.09.1)
 * [Developer] Fixed `SISTR viewer` so plugin and built-in pipelines can use it (20.09.1)
 * [UI] Added two additional fields to SISTR viewer describing the number of alleles found (20.09.1)
+* [Developer]: Updated the session Cart and provided a services for accessing it.
```
[the remark this answers is not in the document]
Comment (joshsadam) [reply CHANGELOG.md:9]: Updated in 413e19783ef142068ca7a6d5a36c8bcfd9d03ce6

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/ajax/dto/cart/CartProject.java (hunk 1 above)
Comment (joshsadam) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/ajax/dto/cart/CartProject.java:10]: Updated in 355003282dbbbbf2cc92247eb2a8b80b5b121a5f

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/cart/CartController.java (hunk 2 above)
Comment (joshsadam) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/cart/CartController.java:25]: Updated in 720b1f646043b2d45e37716fa49ddeb822744838

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/cart/CartGalaxyController.java (hunk 3 above)
Comment (joshsadam) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/cart/CartGalaxyController.java:29]: Updated in 0019e2334d708eab552b72a2f9e3acea265c71c3

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/services/UICartService.java (hunk 7 above)
Comment (joshsadam) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/services/UICartService.java:116]: I left these calls pretty much as they were previous so to not have to refactor too much of the UI calls that are older.  As those get updated we can move these to DTOs.  I think this in particular is a galaxy call.

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/pipelines/PipelineController.java (hunk 6 above)
Comment (joshsadam) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/pipelines/PipelineController.java:105]: This is an old relic, I cannot remember you used to like that.  Removed in 746b003e81aa0d6e9e5f6b072670b1c231c0d8fc

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/sessionAttrs/Cart.java (hunk 8 above)
Comment (joshsadam) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/sessionAttrs/Cart.java:13]: Added in ff327b3dfaf6c5db4002fe5a0c32ffbf4a7512c7

File: src/main/webapp/resources/js/apis/cart/cart.js (hunk 9 above)
Comment (joshsadam) [reply src/main/webapp/resources/js/apis/cart/cart.js:35]: Added in edb5a72614c8656f5bef85e0f90bce70295b0a90

File: src/main/webapp/resources/js/pages/cart/components/CartSamples.jsx (hunk 10 above)
Comment (joshsadam) [reply src/main/webapp/resources/js/pages/cart/components/CartSamples.jsx:83]: Removed in 0b5643b1b98b4b90291a3f3834053255bbd13da7

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/services/UICartService.java (hunk 7 above)
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/services/UICartService.java:115]: Sounds good!

**Review by deepsidhu85 — APPROVED** [review]

**Pull request merged**